Blog post on the support of OpenID4VCI (#682)
keycloak/keycloak-web: blog/2025/issue-credentials-over-openid4vci.adoc

Keycloak implements https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html[OpenID for Verifiable Credential Issuance] (OpenID4VCI). This functionality allows Keycloak to issue verifiable credentials (VCs) as a digital proof of identity or attributes. Enabling this functionality requires consistent configuration across several components, including the realm, client, and issuable credentials (as client scopes).

For example, consider the following scenario: The Keycloak OAuth SIG team wants to issue verifiable membership credentials to its members, including their name and email, which can later be presented at onsite or virtual events as proof of active membership.

NOTE: At the time of writing this blog, Keycloak's support for OpenID4VCI is still experimental. This guide uses the https://www.keycloak.org/nightly/[**nightly release**] which builds from the main branch. The feature is expected to be promoted to preview in early 2026.

== Introduction to OpenID4VCI

> OpenID4VCI is a protocol developed by the OpenID Foundation that extends the OpenID Connect (OIDC) framework to enable the secure and standardized issuance of verifiable credentials (VCs). VCs are digital, tamper-evident representations of information, such as identity attributes or qualifications, that can be cryptographically verified without contacting the issuer at verification time. This promotes decentralized identity management, where users (holders) control their data and share it selectively with verifiers.

Reviewer: I think it's technically not correct to say that OpenID4VCI is built on top of OIDC, instead it is built on top of OAuth, see: https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#section-3.2

That's the reason why it's called OpenID4VCI (OpenID for Verifiable Credential Issuance) and not OpenID Connect 4 Verifiable Credential Issuance.

Reviewer 2: Thank you for spotting that out and providing the correct context and link, it's a great clarification! You are correct that OpenID4VCI is primarily built on top of OAuth 2.0.
I have updated the description accordingly

This will later enable configuring the signing of issued VCs with algorithm `ES256`.

.Screenshot: Adding ECDSA Key Provider
image::${blogImages}/openid4vci/screenshot-realm-ecdsa-generated-provider.png[alt="Screenshot: Adding ECDSA Key Provider", align=center, width=800]

=== Configuring an issuable verifiable credential (as a dedicated client scope)

Different credential types can be configured for issuance depending on what data the credential should embed, what format it should have, or other criteria.
Each credential type is configured as a dedicated client scope of protocol type "OpenID for Verifiable Credentials".

We will configure our membership credential to be issued as an SD-JWT credential carrying a user's first name, last name, and email.
Each claim to be added to the credential requires a corresponding protocol mapper from the User Model to the credential.
Other fields such as the time of issuance or a unique identifier for the credential are supported via other types of protocol mappers.

Since we will be using the Admin REST API to configure the client scope, we first need to obtain a valid Admin Token from the `master` realm to authorize our requests.

> [source,sh]

Reviewer 2: I believe there will be one in the main Keycloak documentation when it’s updated. We also have the [OID4VCI Deployment](https://github.com/keycloak/keycloak-oauth-sig/tree/main/oid4vci-deployment#readme) project, which provides solid examples for both the pre-authorization and authorization code flows and was referenced in the blog post.

:title: Setting Up Keycloak as a Credential Issuer with OpenID4VCI
:date: 2025-12-02
:publish: true
:author: Rodrick Awambeng, Forkim Enjeckayang, Ingrid Kamga, Bertrand Ogen
:summary: Keycloak can act as a verifiable credential issuer using OpenID4VCI, allowing users to receive verifiable credentials as digital proofs of identity. This guide walks through the required configuration steps.

> Before configuring Keycloak, it is helpful to understand its role in decentralized identity ecosystems. As a verifiable credential issuer, Keycloak can issue digitally signed credentials using the https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html[OpenID for Verifiable Credential Issuance] (OpenID4VCI) protocol, allowing relying parties to independently verify them without contacting the issuer.

Reviewer 3: ```suggestion
Before configuring Keycloak, it is helpful to understand its role in decentralized identity ecosystems. As a verifiable credential issuer, Keycloak can issue digitally signed credentials using the https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html[OpenID for Verifiable Credential Issuance] (OpenID4VCI) protocol, allowing relying parties(also known as verifier) to independently verify them without contacting the issuer.
```

.Triangle of trust or Issuer-Holder-Verifier model
image::${blogImages}/openid4vci/triangle-of-trust.png[alt="Triangle of trust or Issuer-Holder-Verifier model", align=center, width=800]

=== Why Use OpenID4VCI

The primary motivations for adopting OpenID4VCI include:

- **Interoperability**: Builds on established OIDC standards, simplifying integration with existing identity providers.
- **Privacy and Security**: Supports selective disclosure (e.g., proving age without revealing birthdate) and offline verification.
- **Compliance**: Aligned with regulations like eIDAS 2.0.
- **Efficiency**: Leverages OIDC mechanisms to streamline issuance while maintaining trust in the "triangle of trust" model (Issuer-Holder-Verifier).

=== Verifiable Credential Formats

> OpenID4VCI supports multiple credential formats, which define how verifiable credentials are encoded and exchanged. *Keycloak currently supports SD-JWT VC for issuance*, while other formats are part of the OpenID4VCI specification.

Reviewer: ```suggestion
OpenID4VCI supports multiple credential formats, which define how verifiable credentials are encoded and issued. *Keycloak currently supports SD-JWT VC for issuance*, while other formats are part of the OpenID4VCI specification.
```

Reviewer: I think "issued" fits the spec terminology better than "exchanged".

- https://drafts.oauth.net/oauth-sd-jwt-vc/draft-ietf-oauth-sd-jwt-vc.html[SD-JWT VC] – Selective Disclosure JSON Web Token Verifiable Credential
- https://identity.foundation/jwt-vc-presentation-profile/[JWT VC] – JSON Web Token Verifiable Credential
- https://www.iso.org/standard/69084.html[mDL/mdoc] – Mobile driver’s license / mobile document format

> These formats allow issuers to provide portable, verifiable digital proofs, supporting selective disclosure and offline verification in decentralized identity ecosystems.

Reviewer: I'd also move the "*Keycloak currently supports SD-JWT VC for issuance*, while other formats are part of the OpenID4VCI specification."-part after this sentense.

=== Key Flows in OpenID4VCI

OpenID4VCI defines two primary flows for credential issuance: the **Authorization Code Flow** and the **Pre-Authorized Code Flow**. These flows determine how a wallet (holder’s application) obtains an access token to request a VC from the issuer.

==== Authorization Code Flow

The Authorization Code Flow is **interactive** and requires the holder to authenticate and consent at the issuer’s authorization endpoint. It is ideal for scenarios where explicit user approval is required, or additional claims must be collected.

**Steps:**

1. The **Wallet** sends an Authorization Request to the **Issuer’s Authorization Endpoint**, requesting credentials.
2. The **Issuer** authenticates the **User** and requests consent for the issuance.
3. The **User** provides credentials and grants consent.
4. The **Issuer** returns an **Authorization Code** to the Wallet.
5. The **Wallet** exchanges the code at the **Issuer Token Endpoint** for an **Access Token**.
6. The **Wallet** requests the **Verifiable Credential** from the **Issuer Credential Endpoint** using the Access Token.
7. The **Issuer** returns the **VC** to the Wallet.

.Authorization Code Flow Diagram

> image::${blogImages}/openid4vci/authorization-code-flow.png[alt="Authorization Code Flow Diagram", align=center, width=800]

Reviewer: ```suggestion
image::${blogImages}/openid4vci/authorization-code-flow.png[alt="Credential issuance via Authorization Code Flow Diagram", align=center, width=800]
```

For example, consider the following scenario: The Keycloak OAuth SIG team wants to issue verifiable membership credentials to its members, including their name and email, which can later be presented at onsite or virtual events as proof of active membership.

NOTE: At the time of writing this blog, Keycloak's support for OpenID4VCI is still experimental. This guide uses the https://www.keycloak.org/nightly/[**nightly release**] which builds from the main branch. The feature is expected to be promoted to preview in early 2026.

== Introduction to OpenID4VCI

OpenID4VCI is a protocol developed by the OpenID Foundation that extends the OAuth 2.0 framework to support the secure and interoperable issuance of Verifiable Credentials (VCs). VCs are digital, tamper-evident representations of information, such as identity attributes or qualifications, that can be cryptographically verified without needing to contact the issuer at verification time. This enables decentralized identity ecosystems in which users (holders) retain control of their data and can share it selectively with verifiers.

.Triangle of trust or Issuer-Holder-Verifier model

> image::${blogImages}/openid4vci/triangle-of-trust.png[alt="Triangle of trust or Issuer-Holder-Verifier model", align=center, width=800]

Reviewer 4: The image's title includes the word "holder" and "verifier", but the image itself does not include these words. How about changing the tile to tell what the image shows? Or adding "verifier" in the image. That might be preferable.

=== Why Use OpenID4VCI

The primary motivations for adopting OpenID4VCI include:

> - **Interoperability**: Builds on established OIDC standards, simplifying integration with existing identity providers.

Reviewer 4: Here the word "OIDC" is shown firstly. So how about showing the official name as follows?
`OpenID Connect (OIDC)`

- **Privacy and Security**: Supports selective disclosure (e.g., proving age without revealing birthdate) and offline verification.
- **Compliance**: Aligned with regulations like eIDAS 2.0.
- **Efficiency**: Leverages OIDC mechanisms to streamline issuance while maintaining trust in the "triangle of trust" model (Issuer-Holder-Verifier).

=== Verifiable Credential Formats

OpenID4VCI supports multiple credential formats, which define how verifiable credentials are encoded and issued.

> - https://drafts.oauth.net/oauth-sd-jwt-vc/draft-ietf-oauth-sd-jwt-vc.html[SD-JWT VC] – Selective Disclosure JSON Web Token Verifiable Credential

Reviewer 4: How about referring IETF official site?
https://www.ietf.org/archive/id/draft-ietf-oauth-sd-jwt-vc-13.html

OpenID4VCI defines two primary flows for credential issuance: the **Authorization Code Flow** and the **Pre-Authorized Code Flow**. These flows determine how a wallet (holder’s application) obtains an access token to request a VC from the issuer.

==== Authorization Code Flow

The Authorization Code Flow is **interactive** and requires the holder to authenticate and consent at the issuer’s authorization endpoint. It is ideal for scenarios where explicit user approval is required, or additional claims must be collected.

**Steps:**

1. The **Wallet** sends an Authorization Request to the **Issuer’s Authorization Endpoint**, requesting credentials.
2. The **Issuer** authenticates the **User** and requests consent for the issuance.
3. The **User** provides credentials and grants consent.
4. The **Issuer** returns an **Authorization Code** to the Wallet.
5. The **Wallet** exchanges the code at the **Issuer Token Endpoint** for an **Access Token**.
6. The **Wallet** requests the **Verifiable Credential** from the **Issuer Credential Endpoint** using the Access Token.
7. The **Issuer** returns the **VC** to the Wallet.

.Credential issuance via Authorization Code Flow

> image::${blogImages}/openid4vci/authorization-code-flow.png[alt="Credential issuance via Authorization Code Flow", align=center, width=800]

Reviewer 4: In this image, to ensure consistency between a request and its response consistent, it might be good add
`Receive Access Token` as a response
after
`Exchange Authorization Code for Access Token (Token Endpont)` as a request

==== Pre-Authorized Code Flow

The Pre-Authorized Code Flow is **non-interactive**. The issuer pre-authenticates and authorizes the user, providing a **pre-authorized code** (often via QR code). It is faster and suitable for pre-approved credential issuance.

**Steps:**

1. The **Issuer** provides a **Pre-Authorized Code** to the **Wallet** (e.g., via QR code or link).
2. The **Wallet** exchanges the code at the **Issuer Token Endpoint** for an **Access Token**.
3. The **Wallet** requests the **Verifiable Credential** from the **Issuer Credential Endpoint** using the Access Token.
4. The **Issuer** returns the **VC** to the Wallet.

.Credential issuance via Pre-Authorized Code Flow

> image::${blogImages}/openid4vci/pre-authorized-code-flow.png[alt="Credential issuance via Pre-Authorized Code Flow", align=center, width=800]

Reviewer 4: In this image, to ensure consistency between a request and its response consistent, it might be good add
`Receive Access Token` as a response
after
`Exchange Authorization Code for Access Token (Token Endpont)` as a request

        "name": "family_name-mapper",
        "protocol": "oid4vc",
        "protocolMapper": "oid4vc-user-attribute-mapper",
        "config": {
          "claim.name": "family_name",
          "userAttribute": "lastName",
          "vc.display": "[{\"name\":\"Family Name\",\"locale\":\"en\"}]"
        }
      },
      {
        "name": "email-mapper",
        "protocol": "oid4vc",
        "protocolMapper": "oid4vc-user-attribute-mapper",
        "config": {
          "claim.name": "email",
          "userAttribute": "email",
          "vc.display": "[{\"name\":\"Email\",\"locale\":\"en\"}]"
        }
      },
      {
        "name": "iat-mapper",
        "protocol": "oid4vc",
        "protocolMapper": "oid4vc-issued-at-time-claim-mapper",
        "config": {
          "claim.name": "iat",
          "truncateToTimeUnit": "HOURS",
          "valueSource": "COMPUTE"
        }
      }
    ]
  }'
----

Commenting on the above configuration:

- Attribute `vc.format` defines the format of the credential whereas `vc.verifiable_credential_type` specifies the `vct` property's value.
- All `vc.display` entries are intended to be used by a wallet to display intelligible descriptions.

> - Known bugs mandate configuring the credential signing algorithm via `vc.proof_signing_alg_values_supported`.

Author: Yes, we are on par. The whole bullet mentioning "Known bugs..." is meant to be removed after ticket https://github.com/keycloak/keycloak/issues/44621 is completed. We hope such updates on the blog post could be done with iterative PRs prior to considering publishing it.

OpenID4VCI defines two primary flows for credential issuance: the **Authorization Code Flow** and the **Pre-Authorized Code Flow**. These flows determine how a wallet (holder’s application) obtains an access token to request a VC from the issuer.

==== Authorization Code Flow

The Authorization Code Flow is **interactive** and requires the holder to authenticate and consent at the issuer’s authorization endpoint. It is ideal for scenarios where explicit user approval is required, or additional claims must be collected.

**Steps:**

1. The **Wallet** sends an Authorization Request to the **Issuer’s Authorization Endpoint**, requesting credentials.
2. The **Issuer** authenticates the **User** and requests consent for the issuance.
3. The **User** provides credentials and grants consent.
4. The **Issuer** returns an **Authorization Code** to the Wallet.
5. The **Wallet** exchanges the code at the **Issuer Token Endpoint** for an **Access Token**.
6. The **Issuer** returns the **Access Token** to the Wallet.
7. The **Wallet** requests the **Verifiable Credential** from the **Issuer Credential Endpoint** using the Access Token.
8. The **Issuer** returns the **VC** to the Wallet.

> .Credential issuance via Authorization Code Flow

Reviewer 5: We can use a Mermaid sequence diagram instead of an image. it will be easier to maintain.
Here is the code for authorization-code-flow.png.

```mermaid
sequenceDiagram
    autonumber
    participant Wallet
    participant User
    participant Issuer as Issuer <br/><Keycloak>

    Wallet->>Issuer: Authorization Request (request credentials)
    Issuer->>User: Authenticate user and request consent
    User->>Issuer: Provide credentials and grant consent
    Issuer-->>Wallet: Return Authorization Code

    Wallet->>Issuer: Exchange Authorization Code for Access Token (Token Endpoint)
    Issuer-->>Wallet: Return Access Token

    Wallet->>Issuer: Request Verifiable Credential (VC Endpoint) using Access Token
    Issuer-->>Wallet: Return Verifiable Credential (VC)
```

image::${blogImages}/openid4vci/authorization-code-flow.png[alt="Credential issuance via Authorization Code Flow", align=center, width=800]

==== Pre-Authorized Code Flow

The Pre-Authorized Code Flow is **non-interactive**. The issuer pre-authenticates and authorizes the user, providing a **pre-authorized code** (often via QR code). It is faster and suitable for pre-approved credential issuance.

**Steps:**

1. The **Issuer** provides a **Pre-Authorized Code** to the **Wallet** (e.g., via QR code or link).
2. The **Wallet** exchanges the code at the **Issuer Token Endpoint** for an **Access Token**.
3. The **Issuer** returns the **Access Token** to the Wallet.
4. The **Wallet** requests the **Verifiable Credential** from the **Issuer Credential Endpoint** using the Access Token.
5. The **Issuer** returns the **VC** to the Wallet.

> .Credential issuance via Pre-Authorized Code Flow

Reviewer 5: Same here. Below is the code for pre-authorization-code-flow.png.

```mermaid
sequenceDiagram
    autonumber
    participant Wallet
    participant Issuer as Issuer <br><Keycloak>

    Issuer-->>Wallet: Send Pre-Authorized Code (QR code or link)

    Wallet->>Issuer: Exchange Pre-Authorized Code for Access Token (Token Endpoint)
    Issuer-->>Wallet: Return Access Token

    Wallet->>Issuer: Request Verifiable Credential (Credential Endpoint) using Access Token
    Issuer-->>Wallet: Return Verifiable Credential (VC)
```

NOTE: At the time of writing this blog, Keycloak's support for OpenID4VCI is still experimental. This guide uses the **main branch snapshot version** `999.0.0-SNAPSHOT`, and the feature is expected to be promoted to preview in early 2026.

== Introduction to OpenID4VCI

OpenID4VCI is a protocol developed by the OpenID Foundation that extends the OpenID Connect (OIDC) framework to enable the secure and standardized issuance of verifiable credentials (VCs). VCs are digital, tamper-evident representations of information, such as identity attributes or qualifications, that can be cryptographically verified without contacting the issuer at verification time. This promotes decentralized identity management, where users (holders) control their data and share it selectively with verifiers.

++++
<div class="paragraph" style="margin-top:1em;margin-bottom:2em;">
    <img src="${blogImages}/openid4vci/triangle-of-trust.png" alt="Triangle of trust or Issuer-Holder-Verifier model" style="border:none;display:block;max-width:100%;height:auto;margin:0 auto;">
    <p style="text-align:center;font-style:italic">Triangle of trust or Issuer-Holder-Verifier model</p>
</div>

> ++++

Reviewer 6: Please use inline HTML only if you need specific formatting / CSS of the images. In all other cases, use the standard AsciiDoc syntax

```
image::... 
```

That even allows for titles of images 

```
.Title
image::... 
```

Let's assume you have already created a realm named `oauth-sig-realm` to represent the OAuth SIG group, and that a couple of users have also been created in this realm, each with a password assigned, to represent members of the group.

In addition, any user who is intended to create credential offers must be granted the `credential-offer-create` role. This ensures that only authorized users can generate credential offers, maintaining the security of the issuance process.

--
++++
<img src="${blogImages}/openid4vci/screenshot-realm-and-users.png" style="margin-bottom:2em" alt="Screenshot: Realm and Users">
++++
--

The behavior of OpenID4VCI across the entire realm can be modulated via realm attributes.
On the Admin Console, find the *OID4VCI Attributes* section under the *Realm Settings > Tokens* tab.

--
++++
<img src="${blogImages}/openid4vci/screenshot-realm-config-oid4vci-attributes.png" style="margin-bottom:2em" alt="Screenshot: OID4VCI Attributes">
++++
--

Sensible defaults apply out of the box, but for the sake of this demo, we would want, for instance, to increase the value of **Pre-Authorized Code Lifespan** to 3 minutes so that the issuance flow expires less quickly.

> To learn more about the **Nonce Lifetime** and other realm attributes for OpenID4VCI not accessible via the Admin Console, please refer to the main https://github.com/keycloak/keycloak/blob/main/docs/documentation/server_admin/topics/oid4vci/vc-issuer-configuration.adoc[Keycloak documentation].

Reviewer 6: The nightly build of the archives and the container also comes with a nightly build of the documentation. So you can link there.

```suggestion
To learn more about the **Nonce Lifetime** and other realm attributes for OpenID4VCI not accessible via the Admin Console, please refer to the main https://www.keycloak.org/docs/nightly/server_admin/index.html#_oid4vci[Keycloak documentation].
```

:title: Setting Up Keycloak as a Credential Issuer with OpenID4VCI

> :date: 2025-12-02

Reviewer 7: Is it possible to update this date for example to `2026-01-16` ? This means Friday this week. I hope we can merge this PR on Friday and hence the date would be accurate (this mechanism of date hardcoded in the PR file itself is not so ideal, but that is how it works in current Keycloak blog posts...).

We covered the necessary configuration steps at the realm, client scope, and client levels, and demonstrated how to retrieve a credential offer to initiate the issuance flow.

If you are looking into streamlining this configuration process for OpenID4VCI in Keycloak, take a look at our https://github.com/keycloak/keycloak-oauth-sig/tree/main/oid4vci-deployment#readme[OID4VCI Deployment] project, which provides solid examples for both the pre-authorization and authorization code flows.
You may also find the https://github.com/keycloak/keycloak-playground/blob/main/fapi-playground/README.md#oid4vci-demo[Keycloak Playground OID4VCI demo] useful as a hands-on demo.

== Feedback & Discussion

We’d love to hear your thoughts on this guide! You can provide feedback or ask questions through:

* Slack: Join the https://slack.cncf.io/[Cloud Native Computing Foundation (CNCF) Slack] and discuss with us in the channel https://cloud-native.slack.com/channels/keycloak-oauth-sig[#keycloak-oauth-sig].

> 

Reviewer 7: ```suggestion
* Github: Inspect and join recent https://github.com/keycloak/keycloak/discussions?discussions_q=is%3Aopen+oid4vci[OpenID4VCI related Github discussions]. Inspect https://github.com/keycloak/keycloak/issues?q=is%3Aissue%20state%3Aopen%20label%3Aarea%2Foid4vc[existing OpenID4VCI related Github issues] and feel free to comment/upvote some of them or create new issue if you have idea for the enhancement or if you found a bug.
```

Is it possible to add a note about feedback on Github? For example something like I've added above? WDYT?

Your input will help us to improve the OpenID4VCI experience in Keycloak.

== Important Note on OpenID4VCI Development Status

> OpenID4VCI support in Keycloak is still under active development. The instructions and configuration options described in this blog post are based on *Keycloak 26.5.0* and may change in future Keycloak versions as the feature evolves.

Reviewer 7: ```suggestion
OpenID4VCI support in Keycloak is still under active development. The instructions and configuration options described in this blog post are based on *Keycloak 26.5.0* and may change in future Keycloak versions as the feature evolves. If you want to look at latest updates in the OID4VCI feature, then you can use latest https://www.keycloak.org/nightly/[Keycloak nightly release], however be aware that instructions in this blog may not work with that version.
```
Is it possible to add some small sentence about "nightly" build as we briefly discussed on today's meeting? For example something like I've suggested?
